routing/unified_policy.go

```diff
@@ -0,0 +1,85 @@
+package routing
+
+import (
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// unifiedPolicy represents a derived policy that covers a set of channels.
+type unifiedPolicy struct {
+	minHTLC lnwire.MilliSatoshi
+	maxHTLC lnwire.MilliSatoshi
+
+	edgePolicy
+}
+
+// update adds a new channel policy to the unified policy. This method can be
+// called on a nil pointer to initialize a unified policy struct. The maxHTLC
+// parameter is the best known upper bound for bandwidth. This can be an
+// accurate estimate, the channel capacity or, when no information is available,
+// lnwire.MaxMilliSatoshi.
+func (u *unifiedPolicy) update(maxHTLC lnwire.MilliSatoshi,
+	policy *channeldb.ChannelEdgePolicy) (
+	*unifiedPolicy, error) {
+
+	// Cap bandwidth by max htlc, if set.
+	if policy.MessageFlags.HasMaxHtlc() &&
+		policy.MaxHTLC < maxHTLC {
+
+		maxHTLC = policy.MaxHTLC
+	}
+
+	// If this is the first channel, copy the policy values into a new
+	// unifiedPolicy struct.
+	if u == nil {
+		ep := edgePolicy{
+			feeBaseMSat:               policy.FeeBaseMSat,
+			feeProportionalMillionths: policy.FeeProportionalMillionths,
+			timeLockDelta:             policy.TimeLockDelta,
+			toNode:                    policy.Node.PubKeyBytes,
+			channelID:                 policy.ChannelID,
+		}
+
+		return &unifiedPolicy{
+			maxHTLC:    maxHTLC,
+			minHTLC:    policy.MinHTLC,
+			edgePolicy: ep,
+		}, nil
+	}
+
+	// Catch bugs that would cause channels from different sets to be
+	// unified.
+	if policy.Node.PubKeyBytes != u.toNode {
+		return nil, errors.New("invalid pubkey")
+	}
+
+	// For fees, set a fee schedule that meets all distinct channel
+	// policies.
+	if policy.FeeBaseMSat > u.feeBaseMSat {
+		u.feeBaseMSat = policy.FeeBaseMSat
+	}
+	if policy.FeeProportionalMillionths > u.feeProportionalMillionths {
+		u.feeProportionalMillionths = policy.FeeProportionalMillionths
+	}
+
+	// For htlc amount constraints, take the widest range.
```

Comment:
are we not trying to take the narrowest range, as to be able to satisfy any of the links? maybe we are thinking about the widths differently

Reply:
I took a different approach where I don't squash all policies together. In stead, the optimal policy is returned based on the actual amount.